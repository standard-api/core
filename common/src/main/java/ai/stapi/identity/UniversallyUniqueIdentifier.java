package ai.stapi.identity;


import java.util.UUID;

public class UniversallyUniqueIdentifier extends UniqueIdentifier {


  protected UniversallyUniqueIdentifier() {
  }

  public UniversallyUniqueIdentifier(String id) {
    super(id);
  }

  public UniversallyUniqueIdentifier(UUID id) {
    super(id.toString());
  }

  public static UniversallyUniqueIdentifier randomUUID() {
    return new UniversallyUniqueIdentifier(UUID.randomUUID());
  }

  public static UniversallyUniqueIdentifier fromString(String id) {
    return new UniversallyUniqueIdentifier(UUID.fromString(id));
  }

  public UUID toUuid() {
    return UUID.fromString(this.getId());
  }
}
