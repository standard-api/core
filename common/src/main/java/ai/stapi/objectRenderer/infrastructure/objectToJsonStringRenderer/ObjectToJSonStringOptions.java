package ai.stapi.objectRenderer.infrastructure.objectToJsonStringRenderer;


import ai.stapi.objectRenderer.model.RendererOptions;
import java.util.Arrays;
import java.util.List;

public class ObjectToJSonStringOptions implements RendererOptions {

  private final List<RenderFeature> features;

  public ObjectToJSonStringOptions(RenderFeature... renderFeatures) {
    this.features = Arrays.stream(renderFeatures).toList();
  }

  public ObjectToJSonStringOptions(List<RenderFeature> renderFeatures) {
    this.features = renderFeatures;
  }

  public List<RenderFeature> getFeatures() {
    return features;
  }

  public enum RenderFeature {
    SORT_FIELDS,
    HIDE_CREATED_AT,
    HIDE_DISPATCHED_AT,
    HIDE_IDS,
    HIDE_KEY_HASHCODE, RENDER_GETTERS
  }
}