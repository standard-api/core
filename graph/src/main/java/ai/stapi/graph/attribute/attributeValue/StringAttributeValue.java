package ai.stapi.graph.attribute.attributeValue;

public class StringAttributeValue extends AbstractAttributeValue<String> implements StringLikeAttributeValue<String> {

  public static final String SERIALIZATION_TYPE = "string";

  public StringAttributeValue(String value) {
    super(value.strip());
  }

  @Override
  public String getDataType() {
    return SERIALIZATION_TYPE;
  }

  @Override
  public String toStringValue() {
    return this.getValue();
  }
}
