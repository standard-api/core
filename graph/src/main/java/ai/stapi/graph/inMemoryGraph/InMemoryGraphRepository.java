package ai.stapi.graph.inMemoryGraph;

import ai.stapi.graph.AttributeContainer;
import ai.stapi.graph.EdgeRepository;
import ai.stapi.graph.EdgeTypeInfo;
import ai.stapi.graph.Graph;
import ai.stapi.graph.NodeIdAndType;
import ai.stapi.graph.NodeInfo;
import ai.stapi.graph.NodeRepository;
import ai.stapi.graph.NodeTypeInfo;
import ai.stapi.graph.RepositoryEdgeLoader;
import ai.stapi.graph.RepositoryNodeLoader;
import ai.stapi.graph.exceptions.GraphElementNotFound;
import ai.stapi.graph.graphElementForRemoval.EdgeForRemoval;
import ai.stapi.graph.graphElementForRemoval.GraphElementForRemoval;
import ai.stapi.graph.graphElementForRemoval.NodeForRemoval;
import ai.stapi.graph.inputGraphElements.InputEdge;
import ai.stapi.graph.inputGraphElements.InputNode;
import ai.stapi.graph.traversableGraphElements.TraversableEdge;
import ai.stapi.graph.traversableGraphElements.TraversableGraphElement;
import ai.stapi.graph.traversableGraphElements.TraversableNode;
import ai.stapi.identity.UniqueIdentifier;
import java.util.List;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentSkipListSet;
import org.jetbrains.annotations.TestOnly;

public class InMemoryGraphRepository implements NodeRepository, EdgeRepository {

  private Graph graph;
  private ConcurrentHashMap<UniqueIdentifier, ConcurrentSkipListSet<TraversableEdge>> nodesEdgesMap;

  public InMemoryGraphRepository() {
    this(new Graph());
  }

  public InMemoryGraphRepository(Graph graph) {
    this.graph = graph;
    this.recalculateNodesEdgesMap();
  }

  public InMemoryGraphRepository(AttributeContainer... inputGraphElements) {
    this(
        new Graph(inputGraphElements)
    );
  }

  @Override
  public void save(InputEdge inputEdge) {
    this.graph = this.graph.with(inputEdge);
    NodeEdgesMapOperation.upsertEdge(this.nodesEdgesMap, this.toTraversableEdge(inputEdge));
  }

  @Override
  public void removeEdge(UniqueIdentifier edgeId, String edgeType) {
    var edgeToRemove = this.graph.getEdge(edgeId, edgeType);
    this.graph = this.graph.removeEdge(
        new EdgeForRemoval(edgeId, edgeType)
    );
    NodeEdgesMapOperation.removeEdge(this.nodesEdgesMap, edgeToRemove);

  }

  @Override
  public void replace(InputEdge inputEdge) {
    var edgeToRemove = this.graph.getEdge(inputEdge.getId(), inputEdge.getType());

    var oldGraph = this.graph;
    this.graph = oldGraph.replace(inputEdge);
    var newEdge = this.loadEdge(inputEdge.getId(), inputEdge.getType());
    NodeEdgesMapOperation.removeEdge(this.nodesEdgesMap, edgeToRemove);
    NodeEdgesMapOperation.upsertEdge(this.nodesEdgesMap, newEdge);
  }

  @Override
  public void removeEdge(EdgeForRemoval edgeForRemoval) {
    this.graph = this.graph.removeEdge(edgeForRemoval);
  }

  @Override
  public void replace(InputNode node) {
    this.graph = this.graph.replace(node);
  }

  @Override
  public void removeNode(UniqueIdentifier id, String nodeType) {
    this.graph = this.graph.removeNode(id, nodeType);
  }

  @Override
  public void removeNode(NodeForRemoval nodeForRemoval) {
    this.graph = this.graph.removeNode(nodeForRemoval);
  }

  public void removeAllElements() {
    this.graph = new Graph();
  }

  private TraversableNode toTraversableNode(InputNode inputNode) {
    return new TraversableNode(
        inputNode.getId(),
        inputNode.getType(),
        inputNode.getVersionedAttributes(),
        new RepositoryEdgeLoader(this)
    );
  }

  private TraversableEdge toTraversableEdge(InputEdge inputEdge) {
    if (inputEdge == null) {
      return null;
    }
    return new TraversableEdge(
        inputEdge.getId(),
        this.loadNode(inputEdge.getNodeFromId()),
        inputEdge.getType(),
        this.loadNode(inputEdge.getNodeToId()),
        inputEdge.getVersionedAttributes(),
        new RepositoryNodeLoader(this)
    );
  }

  @Override
  public TraversableEdge loadEdge(UniqueIdentifier id, String type) {
    return this.toTraversableEdge(
        this.graph.getEdge(id, type)
    );
  }

  public TraversableEdge loadEdge(UniqueIdentifier id) {
    return this.toTraversableEdge(
        this.graph.getEdge(id)
    );
  }

  @Override
  public boolean edgeExists(UniqueIdentifier id, String type) {
    return this.graph.edgeExists(id, type);
  }

  protected boolean edgeExists(UniqueIdentifier id) {
    return this.graph.edgeExists(id);
  }

  @Override
  public List<EdgeTypeInfo> getEdgeTypeInfos() {
    return this.graph.getEdgeTypeInfos();
  }

  public Set<TraversableEdge> findInAndOutEdgesForNode(
      UniqueIdentifier nodeId,
      String nodeType
  ) {

    return NodeEdgesMapOperation.getNodeEdges(this.nodesEdgesMap, nodeId, nodeType);
  }

  @Override
  public TraversableEdge findEdgeByTypeAndNodes(
      String edgeType, 
      NodeIdAndType nodeFrom,
      NodeIdAndType nodeTo
  ) {
    return this.toTraversableEdge(
        this.graph.findEdgeByTypeAndNodes(
            edgeType,
            nodeFrom,
            nodeTo
        )
    );
  }

  @Override
  public void save(InputNode node) {
    this.graph = this.graph.with(node);
  }

  @Override
  public TraversableNode loadNode(UniqueIdentifier uniqueIdentifier, String nodeType) {
    return this.toTraversableNode(
        this.graph.getNode(uniqueIdentifier, nodeType)
    );
  }

  @Override
  public TraversableNode loadNode(UniqueIdentifier uniqueIdentifier) {
    return this.toTraversableNode(
        this.graph.getNode(uniqueIdentifier)
    );
  }

  @Override
  public boolean nodeExists(UniqueIdentifier id, String nodeType) {
    return this.graph.nodeExists(id, nodeType);
  }

  public boolean nodeExists(UniqueIdentifier id) {
    return this.graph.nodeExists(id);
  }

  @Override
  public List<NodeTypeInfo> getNodeTypeInfos() {
    return this.graph.getNodeTypeInfos();
  }

  @Override
  public List<NodeInfo> getNodeInfosBy(String nodeType) {
    return this.graph.getNodeInfosBy(nodeType);
  }

  public List<TraversableNode> loadAllNodes(String nodeType) {
    return this.graph.getAllNodes().stream().filter(
        inputNode -> inputNode.getType().equals(nodeType)
    ).map(this::toTraversableNode).toList();
  }

  public List<TraversableNode> loadAllNodes() {
    return this.graph.getAllNodes().stream()
        .map(this::toTraversableNode).toList();
  }

  public TraversableGraphElement loadGraphElement(UniqueIdentifier uuid) {
    if (this.nodeExists(uuid)) {
      return this.loadNode(uuid);
    }
    if (this.edgeExists(uuid)) {
      return this.loadEdge(uuid);
    }

    throw new GraphElementNotFound(uuid);
  }

  public List<TraversableEdge> loadAllEdges() {
    List<InputEdge> allEdges = this.graph.getAllEdges();
    return allEdges.stream()
        .map(this::toTraversableEdge).toList();
  }

  public List<TraversableEdge> loadAllEdges(String edgeType) {
    return this.graph.getAllEdges().stream()
        .filter(inputEdge -> inputEdge.getType().equals(edgeType))
        .map(this::toTraversableEdge).toList();
  }

  public void merge(Graph otherGraph) {
    this.graph = this.graph.merge(otherGraph);
    this.recalculateNodesEdgesMap();

  }

  public void merge(InMemoryGraphRepository otherGraph) {
    this.merge(otherGraph.getGraph());
  }


  public void removeGraphElements(GraphElementForRemoval... graphElementsForRemoval) {
    this.graph =
        this.graph.removeGraphElements(graphElementsForRemoval);
    this.recalculateNodesEdgesMap();
  }

  public Graph getGraph() {
    return graph;
  }


  public void merge(Graph otherGraph, DeduplicateOptions options) {
    this.graph = this.graph.merge(otherGraph, options);
    this.recalculateNodesEdgesMap();
  }
  
  @TestOnly
  public void prune() {
    this.graph = new Graph();
  }

  private void recalculateNodesEdgesMap() {
    this.nodesEdgesMap = new ConcurrentHashMap<>();
    for (var edge : this.loadAllEdges()) {
      NodeEdgesMapOperation.upsertEdge(this.nodesEdgesMap, edge);
    }
  }
}
