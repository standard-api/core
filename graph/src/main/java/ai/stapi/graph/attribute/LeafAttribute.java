package ai.stapi.graph.attribute;


import ai.stapi.graph.attribute.attributeValue.AttributeValue;
import java.sql.Timestamp;
import java.util.Map;

public class LeafAttribute<T, AV extends AttributeValue<T>> extends AbstractAttribute<T> {

  public static final String DATA_STRUCTURE_TYPE = "LeafAttribute";
  private final AV boxedValue;

  public LeafAttribute(String name, Map<String, MetaData> metaData, AV boxedValue) {
    super(name, metaData);
    this.boxedValue = boxedValue;
  }

  public LeafAttribute(String name, Timestamp createdAt, AV boxedValue) {
    super(name, createdAt);
    this.boxedValue = boxedValue;
  }

  public LeafAttribute(String name, AV boxedValue) {
    super(name);
    this.boxedValue = boxedValue;
  }

  public AV getBoxedValue() {
    return boxedValue;
  }

  @Override
  public T getValue() {
    return this.getBoxedValue().getValue();
  }

  @Override
  public LeafAttribute<T, AV> getCopy() {
    return new LeafAttribute<>(this.getName(), this.getCreatedAt(), this.getBoxedValue());
  }
}
