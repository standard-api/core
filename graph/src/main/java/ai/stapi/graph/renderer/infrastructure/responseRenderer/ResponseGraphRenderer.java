package ai.stapi.graph.renderer.infrastructure.responseRenderer;

import ai.stapi.graph.inMemoryGraph.InMemoryGraphRepository;
import ai.stapi.graph.renderer.infrastructure.responseRenderer.edge.ResponseEdgeRenderer;
import ai.stapi.graph.renderer.infrastructure.responseRenderer.node.ResponseNodeRenderer;
import ai.stapi.graph.renderer.infrastructure.responseRenderer.responseGraph.ResponseEdge;
import ai.stapi.graph.renderer.infrastructure.responseRenderer.responseGraph.ResponseGraph;
import ai.stapi.graph.renderer.infrastructure.responseRenderer.responseGraph.ResponseNode;
import ai.stapi.graph.renderer.model.GraphRenderer;
import ai.stapi.graph.renderer.model.nodeRenderer.RendererOptions;
import java.util.Comparator;
import java.util.LinkedHashMap;


public class ResponseGraphRenderer implements GraphRenderer {

  private final ResponseNodeRenderer responseNodeRenderer;
  private final ResponseEdgeRenderer responseEdgeRenderer;

  public ResponseGraphRenderer(
      ResponseNodeRenderer responseNodeRenderer,
      ResponseEdgeRenderer responseEdgeRenderer
  ) {
    this.responseNodeRenderer = responseNodeRenderer;
    this.responseEdgeRenderer = responseEdgeRenderer;
  }

  @Override
  public boolean supports(RendererOptions options) {
    return options instanceof ResponseRendererOptions;
  }

  @Override
  public ResponseGraph render(InMemoryGraphRepository graph) {
    return this.render(graph, new ResponseRendererOptions());
  }

  @Override
  public ResponseGraph render(InMemoryGraphRepository graph,
      RendererOptions options) {
    var nodes = new LinkedHashMap<String, ResponseNode>();
    var edges = new LinkedHashMap<String, ResponseEdge>();

    graph.loadAllNodes().stream()
        .sorted(Comparator.comparing(
            node -> node.getSortingNameWithNodeTypeFallback() + node.getIdlessHashCodeWithEdges()))
        .map(traversableNode -> this.responseNodeRenderer.render(traversableNode, options))
        .forEach(responseNode -> nodes.put(responseNode.getUuid(), responseNode));

    graph.loadAllEdges().stream()
        .map(traversableEdge -> this.responseEdgeRenderer.render(traversableEdge, options))
        .sorted()
        .forEach(responseEdge -> edges.put(responseEdge.getUuid(), responseEdge));
    return new ResponseGraph(nodes, edges);
  }
}
