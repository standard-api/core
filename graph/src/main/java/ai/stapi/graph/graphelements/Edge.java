package ai.stapi.graph.graphelements;

import ai.stapi.graph.AttributeContainer;
import ai.stapi.graph.NodeIdAndType;
import ai.stapi.graph.attribute.Attribute;
import ai.stapi.graph.inMemoryGraph.exceptions.GraphEdgesCannotBeMerged;
import ai.stapi.graph.traversableGraphElements.TraversableEdge;
import ai.stapi.graph.versionedAttributes.VersionedAttributeGroup;
import ai.stapi.identity.UniqueIdentifier;
import ai.stapi.identity.UniversallyUniqueIdentifier;
import java.util.Objects;

public class Edge extends AbstractGraphElement {

  private final UniqueIdentifier nodeFromId;
  private final UniqueIdentifier nodeToId;
  private final String nodeFromType;
  private final String nodeToType;

  public Edge(
      UniqueIdentifier id,
      String type,
      VersionedAttributeGroup versionedAttributes,
      NodeIdAndType nodeFromIdAndType,
      NodeIdAndType nodeToIdAndType
  ) {
    super(id, type, versionedAttributes);
    this.nodeFromId = nodeFromIdAndType.getId();
    this.nodeToId = nodeToIdAndType.getId();
    this.nodeFromType = nodeFromIdAndType.getType();
    this.nodeToType = nodeToIdAndType.getType();
  }

  public Edge(
      UniqueIdentifier edgeId,
      Node nodeFrom,
      String edgeType,
      Node nodeTo
  ) {
    super(edgeId, edgeType);
    this.nodeFromId = nodeFrom.getId();
    this.nodeFromType = nodeFrom.getType();
    this.nodeToId = nodeTo.getId();
    this.nodeToType = nodeTo.getType();
  }

  public Edge(
      UniqueIdentifier edgeId,
      String edgeType,
      Node nodeFrom,
      Node nodeTo,
      VersionedAttributeGroup attributeGroup
  ) {
    super(edgeId, edgeType, attributeGroup);
    this.nodeFromId = nodeFrom.getId();
    this.nodeFromType = nodeFrom.getType();
    this.nodeToId = nodeTo.getId();
    this.nodeToType = nodeTo.getType();
  }

  public Edge(
      UniqueIdentifier edgeId,
      Node nodeFrom,
      String edgeType,
      Node nodeTo,
      Attribute<?>... attributes
  ) {
    super(edgeId, edgeType, attributes);
    this.nodeFromId = nodeFrom.getId();
    this.nodeFromType = nodeFrom.getType();
    this.nodeToId = nodeTo.getId();
    this.nodeToType = nodeTo.getType();
  }

  public Edge(TraversableEdge traversableEdge) {
    this(
        traversableEdge.getId(),
        traversableEdge.getType(),
        new Node(traversableEdge.getNodeFrom()),
        new Node(traversableEdge.getNodeTo()),
        traversableEdge.getVersionedAttributes()
    );
  }

  public Edge(Node nodeFrom, String edgeType, Node nodeTo) {
    this(
        UniversallyUniqueIdentifier.randomUUID(),
        nodeFrom,
        edgeType,
        nodeTo
    );
  }

  public Edge(
      Node nodeFrom,
      String edgeType,
      Node nodeTo,
      VersionedAttributeGroup attributeGroup
  ) {
    this(UniversallyUniqueIdentifier.randomUUID(), edgeType, nodeFrom, nodeTo,
        attributeGroup);
  }

  public Edge(
      Node nodeFrom,
      String edgeType,
      Node nodeTo,
      Attribute<?>... attributes
  ) {
    this(UniversallyUniqueIdentifier.randomUUID(), nodeFrom, edgeType, nodeTo,
        attributes);
  }

  public Edge(
      UniqueIdentifier edgeId,
      String edgeType,
      UniqueIdentifier nodeFromId,
      UniqueIdentifier nodeToId,
      String nodeFromType,
      String nodeToType,
      VersionedAttributeGroup versionedAttributes
  ) {
    super(edgeId, edgeType, versionedAttributes);
    this.nodeFromId = nodeFromId;
    this.nodeToId = nodeToId;
    this.nodeFromType = nodeFromType;
    this.nodeToType = nodeToType;
  }

  public UniqueIdentifier getNodeFromId() {
    return this.nodeFromId;
  }

  public UniqueIdentifier getNodeToId() {
    return this.nodeToId;
  }

  public String getNodeFromType() {
    return this.nodeFromType;
  }

  public String getNodeToType() {
    return this.nodeToType;
  }

  @Override
  protected AttributeContainer withNewAttributes(VersionedAttributeGroup newAttributes) {
    return new Edge(
        this.getId(),
        this.getType(),
        this.getNodeFromId(),
        this.getNodeToId(),
        this.getNodeFromType(),
        this.getNodeToType(),
        newAttributes
    );
  }

  @Override
  public Edge add(Attribute<?> attribute) {
    return (Edge) super.add(attribute);
  }

  @Override
  protected int getHashCodeWithoutAttributes() {
    return Objects.hash(this.getId()) + this.getIdlessHashCodeWithoutAttributes();
  }

  @Override
  protected int getIdlessHashCodeWithoutAttributes() {
    return Objects.hash(this.getType());
  }

  public NodeIdAndType getNodeFromIdAndType() {
    return new NodeIdAndType(this.nodeFromId, this.nodeFromType);
  }

  public NodeIdAndType getNodeToIdAndType() {
    return new NodeIdAndType(this.nodeToId, this.nodeToType);
  }

  public Edge getCopy() {
    return this;
  }

  public Edge mergeOverwrite(Edge otherEdge) {
    if (!this.getId().equals(otherEdge.getId())) {
      throw GraphEdgesCannotBeMerged.becauseTheyHaveDifferentIds();
    }
    if (!this.getType().equals(otherEdge.getType())) {
      throw GraphEdgesCannotBeMerged.becauseTheyHaveDifferentTypes();
    }
    if (!(this.getNodeFromId().equals(otherEdge.getNodeFromId()))
        || !(this.getNodeToId().equals(otherEdge.getNodeToId()))) {
      throw GraphEdgesCannotBeMerged.becauseTheyHaveDifferentNodeIds();
    }
    return (Edge) this.mergeAttributesWithAttributesOf(otherEdge);
  }
}
