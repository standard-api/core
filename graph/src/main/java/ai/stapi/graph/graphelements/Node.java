package ai.stapi.graph.graphelements;

import ai.stapi.graph.AttributeContainer;
import ai.stapi.graph.attribute.Attribute;
import ai.stapi.graph.inMemoryGraph.exceptions.GraphNodesCannotBeMerged;
import ai.stapi.graph.traversableGraphElements.TraversableNode;
import ai.stapi.graph.versionedAttributes.VersionedAttributeGroup;
import ai.stapi.identity.UniqueIdentifier;
import ai.stapi.identity.UniversallyUniqueIdentifier;
import java.util.Objects;

public class Node extends AbstractGraphElement {


  public Node(UniqueIdentifier id, String nodeType) {
    super(id, nodeType);
  }

  public Node(
      UniqueIdentifier id,
      String type,
      VersionedAttributeGroup versionedAttributes
  ) {
    super(id, type, versionedAttributes);
  }

  public Node(UniqueIdentifier id, String nodeType, Attribute<?>... attributes) {
    super(id, nodeType, attributes);
  }

  public Node(String nodeType) {
    this(UniversallyUniqueIdentifier.randomUUID(), nodeType);
  }

  public Node(TraversableNode node) {
    this(node.getId(), node.getType(), node.getVersionedAttributes());
  }

  public Node(Node node) {
    this(node.getId(), node.getType(), node.getVersionedAttributes());
  }

  public Node(String nodeType, Attribute<?>... attributes) {
    this(UniversallyUniqueIdentifier.randomUUID(), nodeType, attributes);
  }

  @Override
  protected AttributeContainer withNewAttributes(VersionedAttributeGroup newAttributes) {
    return new Node(
        this.getId(),
        this.getType(),
        newAttributes
    );
  }

  @Override
  protected int getHashCodeWithoutAttributes() {
    return Objects.hash(this.getId()) + this.getIdlessHashCodeWithoutAttributes();
  }

  @Override
  protected int getIdlessHashCodeWithoutAttributes() {
    return Objects.hash(this.getType());
  }

  @Override
  public Node add(Attribute<?> attribute) {
    return (Node) super.add(attribute);
  }

  public Node mergeOverwrite(Node otherNode) {
    if (!this.getId().equals(otherNode.getId())) {
      throw GraphNodesCannotBeMerged.becauseTheyHaveDifferentIds();
    }
    if (!this.getType().equals(otherNode.getType())) {
      throw GraphNodesCannotBeMerged.becauseTheyHaveDifferentTypes(
          this.getType(),
          otherNode.getType()
      );
    }
    return (Node) this.mergeAttributesWithAttributesOf(otherNode);
  }

}
