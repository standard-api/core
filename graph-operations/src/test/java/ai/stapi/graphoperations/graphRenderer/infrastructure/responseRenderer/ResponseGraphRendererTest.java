package ai.stapi.graphoperations.graphRenderer.infrastructure.responseRenderer;

import ai.stapi.graph.attribute.LeafAttribute;
import ai.stapi.graph.attribute.attributeValue.StringAttributeValue;
import ai.stapi.graph.inMemoryGraph.InMemoryGraphRepository;
import ai.stapi.graph.graphelements.Edge;
import ai.stapi.graph.graphelements.Node;
import ai.stapi.graph.renderer.infrastructure.responseRenderer.ResponseGraphRenderer;
import ai.stapi.identity.UniversallyUniqueIdentifier;
import ai.stapi.objectRenderer.infrastructure.objectToJsonStringRenderer.ObjectToJSonStringOptions;
import ai.stapi.test.integration.IntegrationTestCase;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;

class ResponseGraphRendererTest extends IntegrationTestCase {

  @Autowired
  private ResponseGraphRenderer responseGraphRenderer;

  @Test
  void itShouldRenderEmptyGraph() {
    var actual = this.responseGraphRenderer.render(new InMemoryGraphRepository());
    this.thenObjectApproved(actual, getOptions());
  }

  @Test
  void itShouldRenderGraphWithNode() {
    var graph = new InMemoryGraphRepository(
        new Node(UniversallyUniqueIdentifier.randomUUID(), "example_node_type")
    );
    var actual = this.responseGraphRenderer.render(graph);
    this.thenObjectApproved(actual, getOptions());
  }

  @Test
  void itShouldRenderGraphWithNodeWithAttribute() {
    var graph = new InMemoryGraphRepository(
        new Node(
            UniversallyUniqueIdentifier.randomUUID(),
            "example_node_type",
            new LeafAttribute<>("exampleAttributeName",
                new StringAttributeValue("exampleAttributeValue"))
        )
    );
    var actual = this.responseGraphRenderer.render(graph);
    this.thenObjectApproved(actual, getOptions());
  }

  @Test
  void itShouldRenderGraphWithEdge() {
    var nodeFrom = new Node(
        UniversallyUniqueIdentifier.fromString("c023fdc9-4045-4642-9f02-f301833ac17c"),
        "example_node_from_type");
    var nodeTo = new Node(
        UniversallyUniqueIdentifier.fromString("0fe04b45-8275-4d30-ae2f-2fc24cef530a"),
        "example_node_to_type");
    var graph = new InMemoryGraphRepository(
        nodeFrom,
        nodeTo,
        new Edge(
            nodeFrom,
            "example_edge_type",
            nodeTo,
            new LeafAttribute<>("exampleAttributeName",
                new StringAttributeValue("exampleAttributeValue"))
        )
    );
    var actual = this.responseGraphRenderer.render(graph);
    this.thenObjectApproved(actual, getOptions());
  }

  private ObjectToJSonStringOptions getOptions() {
    return new ObjectToJSonStringOptions(
        ObjectToJSonStringOptions.RenderFeature.SORT_FIELDS,
        ObjectToJSonStringOptions.RenderFeature.HIDE_CREATED_AT,
        ObjectToJSonStringOptions.RenderFeature.HIDE_IDS
    );
  }
}