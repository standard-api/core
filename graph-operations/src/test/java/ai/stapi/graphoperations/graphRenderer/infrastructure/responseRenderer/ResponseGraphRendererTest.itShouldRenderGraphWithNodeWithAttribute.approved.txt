{
  "edges" : { },
  "nodes" : {
    "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxx" : {
      "type" : "example_node_type",
      "uuid" : "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxx",
      "attributes" : {
        "exampleAttributeName" : {
          "name" : "exampleAttributeName",
          "values" : [ {
            "createdAt" : hidden_date,
            "value" : "exampleAttributeValue"
          } ]
        }
      }
    }
  }
}